### PR TITLE
Add build info for libraries

### PR DESCRIPTION
**Pull request by partouf**
Addresses part of https://github.com/compiler-explorer/compiler-explorer/issues/4446

Adds new button next to search result. When you select a version you can click the button to see what binaries are available for that version for the current compiler.

image

:heavy_check_mark: For clearness it should probably also include the compiler name, but not sure how to get that.

### CODE REVIEW

File: static/widgets/libs-widget.ts [hunk 1]
```diff
@@ -41,6 +41,76 @@ type FavLibraries = Record<string, string[]>;
...
+            const details = await this.get(library, version);
+            for (const detail of details) {
+                if (detail.buildinfo.compilerId === compilerId) {
+                    result.push(detail);
+                } else if (detail.buildinfo.compilerId === 'cshared') {
+                    result.push(detail);
+                }
+            }
+        } catch (e) {
+            // ignore?
```
Comment (mattgodbolt) [inline static/widgets/libs-widget.ts:93]: Perhaps log through Sentry?

File: static/widgets/libs-widget.ts [hunk 2]
```diff
@@ -296,6 +375,44 @@ export class LibsWidget {
...
+    async getBuildInfoAsHtml(libId: string, semver: string, url?: string): Promise<string> {
+        const details = await lib_annotations.getForCompiler(libId, semver, this.currentCompilerId);
+
+        let libInfoText = '';
+        for (const info of details) {
+            if (info.annotation.commithash) {
+                const machineName = shortenMachineName(info.annotation.machine || '');
+                const stdlib = info.buildinfo.libcxx;
+                if (url && url.startsWith('https://github.com/')) {
+                    // this is a bit of a hack because we don't store the git repo in our properties files
```
Comment (mattgodbolt) [inline static/widgets/libs-widget.ts:387]: lol not _that_ much of a hack compared to so many other parts of our code, but thanks for the honest comment :)

File: static/widgets/libs-widget.ts [hunk 3]
```diff
@@ -41,6 +41,76 @@ type FavLibraries = Record<string, string[]>;
...
+};
+
+class LibraryAnnotations {
+    private all: Record<string, LibraryAnnotationDetail[]> = {};
+    constructor() {}
+
+    private async get(library: string, version: string): Promise<LibraryAnnotationDetail[]> {
+        const libver = `${library}/${version}`;
+        if (!Object.keys(this.all).includes(libver)) {
+            const response = await fetch(`https://conan.compiler-explorer.com/annotations/${libver}`);
```
Comment (mattgodbolt) [inline static/widgets/libs-widget.ts:74]: Is this the first time we'll be hitting `conan.compiler-explorer.com` externally? Presuming cloudfront caching is enough? or do we need to stand up somethign a bit more substantial to handle this? (we could cache ourselves in a lambda if it became an issue?) (obviously this caches here locally too)

**Review by mattgodbolt — COMMENTED** [review]
Thanks! This seems reasonable to me, a few minor questions only (but none blocking)

File: static/widgets/libs-widget.ts (hunk 3 above)
Comment (partouf) [reply static/widgets/libs-widget.ts:74]: First time indeed of frontend knowing about conan.compiler-explorer.com, also cloud-front cache is currently disabled, and thats ok in this case. People tend to not ctrl+r refresh until really needed, and that might restrain the IO, though the request itself or library and version is quite intensive. I think we'll have to monitor the situation.

**Comment by partouf** [comment]
live

**Pull request merged**